src/ln/channelmanager.rs

```diff
@@ -442,6 +442,43 @@ macro_rules! try_chan_entry {
 	}
 }
 
+macro_rules! return_monitor_err {
+	($self: expr, $err: expr, $channel_state: expr, $entry: expr, $action_type: path) => {
+		return_monitor_err!($self, $err, $channel_state, $entry, $action_type, Vec::new(), Vec::new())
+	};
+	($self: expr, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $raa_first_dropped_cs: expr) => {
+		if $action_type != RAACommitmentOrder::RevokeAndACKFirst { panic!("Bad return_monitor_err call!"); }
+		return_monitor_err!($self, $err, $channel_state, $entry, $action_type, Vec::new(), Vec::new(), $raa_first_dropped_cs)
+	};
+	($self: expr, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $failed_forwards: expr, $failed_fails: expr) => {
+		return_monitor_err!($self, $err, $channel_state, $entry, $action_type, $failed_forwards, $failed_fails, false)
+	};
+	($self: expr, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $failed_forwards: expr, $failed_fails: expr, $raa_first_dropped_cs: expr) => {
+		match $err {
+			ChannelMonitorUpdateErr::PermanentFailure => {
+				let (channel_id, mut chan) = $entry.remove_entry();
+				if let Some(short_id) = chan.get_short_channel_id() {
+					$channel_state.short_to_id.remove(&short_id);
+				}
+				// TODO: $failed_fails is dropped here, which will cause other channels to hit the
+				// chain in a confused state! We need to move them into the ChannelMonitor which
+				// will be responsible for failing backwards once things confirm on-chain.
+				// It's ok that we drop $failed_forwards here - at this point we'd rather they
+				// broadcast HTLC-Timeout and pay the associated fees to get their funds back than
+				// us bother trying to claim it just to forward on to another peer. If we're
+				// splitting hairs we'd prefer to claim things that went *to us*, but we haven't
+				// given up the preimage yet, so might as well just wait until the payment is
```

Comment:
But we may lost a HTLC routed through us there ? Because nothing guarantee us that peer will choose us again to route its payment

Reply:
I'm not sure what your point is here? We will, indeed, forget about a payment which was in-flight through us, but only for an HTLC that we hadn't yet forwarded onwards, so its not really our problem to track it. Ie in the case of A -> B -> C, when we're B, and we receive the final RAA from A which is supposed to allow us to forward the payment to C, but we fail a monitor update and force-close on chain, we don't really care about tracking the HTLC. We never forwarded it onwards to C, so we have no way to learn the preimage, so there is no way we can claim the payment for ourselves, so we just let A figure it out.

Reply:
Ah no wasn't talking about internal HTLC tracking, I ticked on last sentence of comment "If we're splitting...", which say that payment may be retried (by origin node) but seems to me if we are unreliable doesn't seem that we're gonna be again in payment route and we won't see this payment a second time. But well doesn't seems we can do better if our Monitor is dead (sending back update_fail is useless, if we can't revoke).

Reading again, new point, why speaking about "we'd prefer to claim things that went to us" ? Seems to me not doable even with more complexity as we never forwarded HTLC.

Reply:
I kinda implicitly assume there that the payment went through multiple hops and we only failed one of our channels. So if the sender wants to pay us they'll retry via a different channel. I was referring to payments that *we received* is our own payments. We can't really claim it cause we never generated a PaymentReceived event which means we never got the preimage from the user, but in theory we could do that, use the preimage to claim the now-on-chain HTLC and take the payment, but thats complicated and probably not for much gain.

Reply:
I clarified the comment a bit, hopefully.